source/agora/consensus/protocol/Nominator.d

```diff
@@ -800,6 +800,8 @@ extern(D):
             auto data = deserializeFull!ConsensusData(value[]);
             if (auto fail_reason = this.ledger.validateConsensusData(data))
             {
+                if (fail_reason == Ledger.InvalidConsensusDataReason.MayBeValid)
+                    return ValidationLevel.kInvalidValue;
```

Comment:
By SCP's definition if you return `kInvalidValue` then it means the node may never vote this value as valid afterwards. Are you sure we're not supposed to use `kMaybeValidValue`?

Reply:
Well, initially we returned `kMaybeValidValue` here. But I had some problems in the balloting phase. Node started to confirm statements without having access to all the previous blocks. I completely went from these [two](https://github.com/bpfkorea/agora/blob/bbf523199b6efd1667ee0e3d79ad59d4c3ea93c2/source/agora/consensus/protocol/Nominator.d#L539) [comments](https://github.com/bpfkorea/agora/blob/406a22df4c33dfd2f2e3a61811fe6e73aea94852/source/scpp/src/scp/SCPDriver.h#L56-L58). and decided `kInvalidValue` was more appropriate. In nominating phase it does not matter what we return they are the same. In balloting we should still fully validate the value, but our ledger may be a bit behind to return `kMaybeValidValue`. That is indeed what Stellar does actually. They only return `kMaybeValidValue` [here](https://github.com/stellar/stellar-core/blob/54b03f755ae5d5aa12a799c8f1ee4d87fc9d1a1d/src/herder/HerderSCPDriver.cpp#L237).

Reply:
> then it means the node may never vote this value as valid afterwards

I am not entirely sure, but if the same value is renominated. It should probably count as a new value.

Reply:
Well okay. The docs are not amazing..

Reply:
I mean the SCP docs.